### ISSUE TITLE
Consider raising exception when reading on a closed connection

### ISSUE THREAD

**Issue by aaugustin** [aaugustin/websockets#77]
The pattern for reading a message robustly is:

    message = yield from websocket.recv()
    if message is None:
        break

If a user fails to check for `None` they'll most likely get a confusing error as soon as they try to do something with the message.

In hindsight it would make more sense to raise an exception in that case: the action attempted by the programmer failed.

For comparison the pattern for writing a message robustly is:

    if not websocket.open:
        break
    yield from websocket.send(message)

If a user fails to check for `websocket.open` they get an exception when calling `send()` which makes complete sense.

There are two possibilities:

- change the default behavior of `recv()` to raise an exception on a closed connection and provide a flag to opt-in to the historical behavior: that's best if we believe the new behavior makes more sense
- provide a flag to opt-in to the alternative behavior of `recv()`

We could also have two API, one that returns `None` when the connection is closed and one that raises an exception, but I'm afraid the nuance could be confusing.

**Comment by aaugustin** [comment]
The suggested new behavior wouldn't have confused the reporter of #75.

**Comment by andrewyoung1991** [comment]
+1 on raising errors on closed connections.

it would also be nice to use pythons context manager syntax to guarantee client code can be robust without any need for checking on their end.

something like this:
```python
def echo_handler(websocket):
    with websocket as ws:
        # as long as the socket is open do your thing
        while True:
            message = yield from ws.recv()
            yield from ws.send(message)
```
it's not necessarily pretty (it took 3 indents to write anything useful), but it'll take some responsibility off a client.

**Comment by aaugustin** [comment]
It dosn't matter if the handler ends with an exception; the server doesn't do anything with it.

**Comment by SzieberthAdam** [comment]
I would also prefer getting an exception there.

**Comment by dzen** [comment]
I think the send() method should raise too.
 in `write_frame`  ( https://github.com/aaugustin/websockets/blob/master/websockets/protocol.py#L261 ) the exception `ConnectionResetError` is caught but  the parent code is not warned about the data loss.

**Comment by aaugustin** [comment]
At this point, it's mostly a matter of me feeling in the mood for writing the patch :-)

**Comment by aaugustin** [comment]
@dzen There may be an arbitrary amount of data in-flight between your endpoint and the remote endpoint.

If `send()` hits a connection error while trying to flush the data, it's almost guaranteed that you lost the last frame. You may also have lost arbitrarily many previous frames.

If you want to check if a frame has been received, you need to get that information from the remote endpoint. The easiest way to do so is to `yield from self.ping()`.

**Comment by aaugustin** [comment]
I just committed this change and I plan to release version 3.0 in the next few days.

Folks who are following this issue, if you could try the master branch and report your experience, that would be very useful!

**Comment by aaugustin** [comment]
@dzen Upon further thought, I implemented your suggestion in f531d696dad7a0b3c8fe59e3512992f3bc9c3037. Thanks for bringing that up!
